File: android/core/protocol/src/main/kotlin/app/hapi/protocol/wire/AgentState.kt
package app.hapi.protocol.wire

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull

/**
 * Session agent state (`AgentStateSchema`, `shared/src/schemas.ts`). Pending
 * permission / input requests live here — they are NOT chat messages
 * (`docs/api/client-contract/rest.md`, plan "核心协议事实").
 *
 * Closed zod enums (`startingMode`, request `status`/`decision`) are kept as
 * raw strings on the wire so a hub that grows a new value can never make an
 * entire `Session` undecodable; interpret them via `app.hapi.protocol.catalog`.
 */
@Serializable
data class AgentState(
    val controlledByUser: Boolean? = null,
    /** `'local' | 'remote' | 'pty'` — mode the session was started in. */
    val startingMode: String? = null,
    /** Pending requests keyed by request id (the `:rid` in the approve/deny routes). */
    val requests: Map<String, AgentStateRequest>? = null,
    /** Resolved requests keyed by request id. */
    val completedRequests: Map<String, AgentStateCompletedRequest>? = null,
)

/** A pending tool-permission / user-input request (`AgentStateRequestSchema`). */
@Serializable
data class AgentStateRequest(
    val tool: String,
    /** Tool arguments, shape depends on [tool] (zod `unknown`). */
    val arguments: JsonElement = JsonNull,
    val createdAt: Long? = null,
)

/** A resolved request (`AgentStateCompletedRequestSchema`). */
@Serializable
data class AgentStateCompletedRequest(
    val tool: String,
    val arguments: JsonElement = JsonNull,
    val createdAt: Long? = null,
    val completedAt: Long? = null,
    /** `'canceled' | 'denied' | 'approved'`. */
    val status: String,
    val reason: String? = null,
    val mode: String? = null,
    /** `'approved' | 'approved_for_session' | 'denied' | 'abort'`. */
    val decision: String? = null,
    val allowTools: List<String>? = null,
    /**
     * Answers to AskUserQuestion / request_user_input. Two wire formats:
     * flat `{question: string[]}` and nested `{question: {answers: string[]}}` —
     * kept raw until the permission-UX milestone needs them.
     */
    val answers: JsonElement? = null,
)

File: ios/Packages/HapiKit/Sources/HapiProtocol/Models/AgentState.swift
import Foundation

/// A pending tool-approval / user-input request raised by the agent.
///
/// Mirrors `AgentStateRequestSchema` (`shared/src/schemas.ts`). Permission
/// requests are NOT messages: they live on `Session.agentState.requests`
/// keyed by request id and are resolved over
/// `POST /api/sessions/:id/permissions/:requestId/approve|deny`.
public struct AgentStateRequest: Codable, Equatable, Sendable {
    public var tool: String
    /// Free-form tool arguments (`z.unknown()` on the wire).
    public var arguments: JSONValue?
    /// Epoch ms when the request was raised; older CLIs omit it.
    public var createdAt: Int?

    public init(tool: String, arguments: JSONValue? = nil, createdAt: Int? = nil) {
        self.tool = tool
        self.arguments = arguments
        self.createdAt = createdAt
    }
}

/// A resolved permission request, moved from `requests` to
/// `completedRequests` once decided.
///
/// Mirrors `AgentStateCompletedRequestSchema` (`shared/src/schemas.ts`).
public struct AgentStateCompletedRequest: Codable, Equatable, Sendable {
    public enum Status: String, Codable, Sendable {
        case canceled
        case denied
        case approved
    }

    public enum Decision: String, Codable, Sendable {
        case approved
        case approvedForSession = "approved_for_session"
        case denied
        case abort
    }

    public var tool: String
    public var arguments: JSONValue?
    public var createdAt: Int?
    public var completedAt: Int?
    public var status: Status
    public var reason: String?
    /// Permission mode switched to while approving, when any.
    public var mode: String?
    public var decision: Decision?
    public var allowTools: [String]?
    /// Two wire formats, kept raw:
    /// flat `{question: [answers]}` (AskUserQuestion) or
    /// nested `{question: {answers: [...]}}` (request_user_input).
    public var answers: JSONValue?

    public init(
        tool: String,
        arguments: JSONValue? = nil,
        createdAt: Int? = nil,
        completedAt: Int? = nil,
        status: Status,
        reason: String? = nil,
        mode: String? = nil,
        decision: Decision? = nil,
        allowTools: [String]? = nil,
        answers: JSONValue? = nil
    ) {
        self.tool = tool
        self.arguments = arguments
        self.createdAt = createdAt
        self.completedAt = completedAt
        self.status = status
        self.reason = reason
        self.mode = mode
        self.decision = decision
        self.allowTools = allowTools
        self.answers = answers
    }
}

/// Agent-owned session state carried on `Session.agentState` and patched over
/// SSE under the `agentStateVersion` gate.
///
/// Mirrors `AgentStateSchema` (`shared/src/schemas.ts:195-203`).
public struct AgentState: Codable, Equatable, Sendable {
    /// `true` while the terminal owns the session (remote control paused).
    public var controlledByUser: Bool?
    /// Launch mode the session was started in (persisted for reopen/resume).
    public var startingMode: SessionStartingMode?
    /// Pending requests keyed by request id.
    public var requests: [String: AgentStateRequest]?
    /// Resolved requests keyed by request id.
    public var completedRequests: [String: AgentStateCompletedRequest]?

    public init(
        controlledByUser: Bool? = nil,
        startingMode: SessionStartingMode? = nil,
        requests: [String: AgentStateRequest]? = nil,
        completedRequests: [String: AgentStateCompletedRequest]? = nil
    ) {
        self.controlledByUser = controlledByUser
        self.startingMode = startingMode
        self.requests = requests
        self.completedRequests = completedRequests
    }
}
